File: PhixClr/PChar.cpp

#include "stdafx.h"
#include "Phix.h"
#include "Interpreter.h"

template <typename T, unsigned S>
inline unsigned arraysize(const T(&v)[S]) { return S; }

bool PChar::IsDigit() {
	switch (Value) {
	case '1':
	case '2':
	case '3':
	case '4':
	case '5':
	case '6':
	case '7':
	case '8':
	case '9':
	case '0':
		return true;
	}
	return false;
}

bool PChar::IsWhiteSpace() {
	return Value == ' ' || Value == '\n' || Value == '\r' || Value == '\t';
}

bool PChar::IsOperator() {

	int asize = arraysize(Operators);
	for (int i = 0; i < asize; i++) {
		if (Operators[i] == Value) return true;
	}
	return false;
}

File: Phix/PChar.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Phix {
	class PChar {
		
		public char Value { get; private set; }

		public static implicit operator char(PChar p) {
			return p.Value;
		}

		public static implicit operator PChar(char c) {
			return new PChar(c);
		}

		public bool IsWhiteSpace() {
			return Value == ' ' || Value == '\n' || Value == '\r' || Value == '\t';
		}

		public bool IsDigit() {
			return Value.IsDigit();
		}

		public bool IsOperator() {
			return Lexer.Operators.Contains(Value);
		}

		public override bool Equals(object obj) {
			return Value.Equals(obj);
		}

		public override string ToString() {
			return Value.ToString();
		}

		public override int GetHashCode() {
			return Value.GetHashCode();
		}

		public PChar(char c) {
			Value = c;
		}
	}
}
